src/v/raft/replicate_batcher.cc

```diff
@@ -117,17 +116,35 @@ ss::future<> replicate_batcher::flush() {
                 // this problem caused truncation failure.
 
                 if (!_ptr->is_leader()) {
-                    for (auto& n : notifications) {
+                    for (auto& n : item_cache) {
                         n->_promise.set_value(errc::not_leader);
                     }
                     return ss::make_ready_future<>();
                 }
 
                 auto meta = _ptr->meta();
                 auto const term = model::term_id(meta.term);
-                for (auto& b : data) {
-                    b.set_term(term);
+                ss::circular_buffer<model::record_batch> data;
+                std::vector<item_ptr> notifications;
+
+                for (auto& n : item_cache) {
+                    if (
```

Comment:
before this flush operation was all or nothing; if leader flush everything, else nothing. now that some cached batches may be conditionally dropped, the concern would be that some other state exists (e.g. in the batcher) that assumes this behavior. maybe a counter something. i don't have anything in mind, and you probably already thought of this, it is just the only thing that comes to mind as a concern.

Reply:
I've double checked batcher and the up-stack (consensus) and haven't found any assumptions about the continuity of the item_cache.